# Taller MCDA para e-commerce
**Decisiones multicriterio axiológicamente fundamentadas** – E-commerce, TAM, tecnología persuasiva, ética, credibilidad y optimización con PuLP.

Nombre: ____________________ &nbsp;&nbsp; Fecha: ____________

En este notebook se desarrolla la guía en el mismo orden en que aparece: primero las actividades conceptuales (1 a 7), luego el modelo en Python con PuLP (pasos 9 a 19, con las actividades 8 a 11) y finalmente la conclusión, el desafío final y las preguntas de cierre.

## Paso 1 – Racionalidad limitada

### Actividad 1 – Identificar límites de conocimiento

**Escriba tres aspectos que la organización no conoce con certeza y qué evidencia necesitaría para reducir cada incertidumbre.**

| Incertidumbre | Por qué no se sabe | Evidencia para reducirla |
|---|---|---|
| Costos de mantenimiento a 3 años | El TCO es una estimación; pueden subir licencias, soporte o integraciones | Cotizaciones formales y contratos del proveedor, experiencia de otras empresas que usen la plataforma |
| Comportamiento con más demanda o carga | La latencia se midió en un momento, no en peaks (ej. Cyber Day) | Pruebas de carga y estrés, SLA del proveedor, historial de disponibilidad |
| Impacto real de una técnica persuasiva sobre la conversión | No sabemos si la escasez o la prueba social aumentan las ventas con nuestros clientes, ni si generan desconfianza | Pruebas A/B en un piloto, tasa de conversión y de reclamos/devoluciones |

También es incierto el comportamiento futuro de los usuarios, es decir, si la aceptación medida en una prueba se mantiene en el uso diario.

**Pregunta epistemológica: ¿qué sabemos realmente?**

- **Dato medido (observado):** TCO estimado, latencia medida (ej. 250 ms en ShopFast), disponibilidad, promedios PU/PEOU de una encuesta.
- **Suposición (expectativa):** que esa latencia se mantenga con más carga, que el TCO no suba en 3 años, que los usuarios encuestados representen a todos los clientes.

## Paso 2 – MCDA como apoyo a la decisión

### Actividad 2 – Formular la decisión

> "Dadas las alternativas **ShopFast, TrustCart, LeanCommerce y ProCommerce**, la organización debe seleccionar **una única plataforma de e-commerce** considerando **valor económico, aceptación de los usuarios, desempeño tecnológico, uso responsable de la persuasión y cumplimiento ético/jurídico**."

- **¿Quién decide?** La gerencia de la organización (por ejemplo, la gerencia comercial junto con la gerencia de TI).
- **¿Quién utilizará la plataforma?** Los clientes que compran y los trabajadores internos que la administran (ventas, marketing, TI).
- **¿Quién puede ser afectado por la decisión?** Los clientes (sus datos y su autonomía al comprar), la organización (costos e ingresos), el equipo TI (mantenimiento) y la sociedad en general.
- **¿Quién establece restricciones?** El área legal/compliance, el regulador (SERNAC, Ley 19.496, Decreto 6 de 2021) y las políticas internas y código de ética de la organización.

## Paso 3 – Comenzar por valores, no por pesos

Valor superior: **V0 = Adopción responsable y sostenible de una solución de e-commerce**, con las dimensiones V1 Valor para el usuario, V2 Valor económico, V3 Responsabilidad ética/jurídica y V4 Sostenibilidad tecnológica.

### Actividad 3 – Construir el árbol axiológico

| Stakeholder | ¿Qué valora? | ¿Qué objetivo podría representarlo? | Dimensión |
|---|---|---|---|
| Cliente | Comprar fácil y rápido, que no lo engañen, que cuiden sus datos | Maximizar facilidad de uso (PEOU) y utilidad (PU); no usar persuasión engañosa | V1 |
| Organización | Que la plataforma sea rentable y aumente las ventas | Minimizar TCO; maximizar soporte de persuasión responsable | V2 |
| Equipo tecnológico | Que sea estable, rápida y fácil de mantener | Maximizar disponibilidad y mantenibilidad, minimizar latencia | V4 |
| Área legal/compliance | Cumplir la ley y las políticas internas | Cumplir Ley 19.496, Decreto 6/2021 y normativa de datos personales | V3 |
| Sociedad/regulador | Comercio electrónico justo y seguro | Respetar la autonomía del consumidor y la seguridad de la información | V3 |

De aquí salen los criterios del modelo: los objetivos de V1, V2 y V4 quedan como criterios con peso (compensatorios) y los de V3 quedan como gates (no compensatorios).

## Paso 4 – Aceptación tecnológica con TAM

### Actividad 4 – Distinguir dato y constructo

**¿Por qué no sería suficiente escribir directamente `tam = {"A": 9, "B": 8, "C": 7, "D": 9}`?**

- No se sabe de dónde salen esos números: quién los puso, si se midió algo o si es la opinión del equipo.
- TAM tiene dos constructos distintos (PU y PEOU) y ahí se juntan en un solo número, por lo que se pierde información (una plataforma puede ser útil pero difícil de usar).
- Según Davis, PU y PEOU se miden con ítems tipo Likert respondidos por usuarios, no se inventan. Un 9 sin instrumento es un número computable, pero no creíble.

**¿Qué información adicional necesitaría para considerar esos valores creíbles?**

- El cuestionario usado (ítems PU1–PU3 y PEOU1–PEOU3) y la escala (Likert 1 a 5).
- Cuántas personas respondieron y si eran usuarios representativos (clientes reales, no solo gente del equipo).
- La tarea que realizaron antes de responder (ej. una compra de prueba en cada plataforma).
- Fecha de la medición, quién la hizo, y el promedio y la dispersión de cada constructo.

## Paso 5 – Tecnología persuasiva, autonomía y valores

### Actividad 5 – Auditar una técnica persuasiva

Técnica seleccionada: **escasez** ("Quedan 3 unidades").

| Pregunta | Respuesta |
|---|---|
| ¿Qué conducta intenta facilitar? | Que el cliente decida comprar más rápido y no deje el producto para después. |
| ¿Qué evidencia muestra al usuario? | El stock disponible. Para que sea responsable, el número debe venir del inventario real en ese momento. |
| ¿Puede ser rechazada? | Sí, el usuario puede ignorar el mensaje y no comprar; no lo bloquea ni le pone contadores de tiempo falsos. |
| ¿Conserva la autonomía? | Sí, siempre que el dato sea verdadero: el usuario recibe información real y decide él. |
| ¿Qué la convertiría en manipulación? | Mostrar "quedan 3" cuando hay 742 (falsa escasez), temporizadores que se reinician o esconder que el mensaje es automático. Eso es influencia + engaño. |

Con el criterio del taller (persuasión responsable = influencia + veracidad + autonomía + transparencia), la escasez es responsable solo si el stock es real y el usuario puede ignorarla. Este tipo de prácticas es lo que se evalúa en el gate de autonomía.

## Paso 6 – Privacidad y contexto

### Actividad 6 – Evaluar un flujo de datos ("Productos para ti")

Se usa el marco de integridad contextual de Nissenbaum: quién + qué dato + para qué + con quién + bajo qué norma.

1. **¿Qué datos utiliza?** Historial de compras, productos vistos, búsquedas, carrito y posiblemente ubicación o datos del perfil.
2. **¿Quién los recolecta?** La tienda a través de la plataforma de e-commerce; también puede intervenir el proveedor de la plataforma o un servicio externo de recomendaciones.
3. **¿Cuál es la finalidad?** Mostrar productos que podrían interesarle al cliente, para facilitar la compra y aumentar las ventas.
4. **¿Quién recibe la inferencia?** El mismo cliente (ve las recomendaciones) y la tienda. No debería llegar a terceros (ej. vender el perfil a otras empresas).
5. **¿El uso es esperable en el contexto?** Sí, en una tienda online es normal que recomienden productos según lo que compré o vi en esa misma tienda. No sería esperable que usen datos de otros sitios, datos sensibles (ej. salud) o que compartan el perfil con terceros.
6. **¿Qué control posee el usuario?** Debería poder ver por qué se le recomienda algo, desactivar la personalización y borrar su historial, además de estar informado en la política de privacidad.

**Norma aplicable:** Ley 19.628 (vigente) y Ley 21.719, que entra en vigencia el 1 de diciembre de 2026. Como el taller se realiza antes de esa fecha, la 21.719 se considera normativa futura, pero conviene que la plataforma ya se ajuste a ella.

**Conclusión:** el flujo respeta la integridad contextual si los datos se quedan dentro de la tienda y se usan solo para recomendar. Si se comparten con terceros o se usan para otra finalidad, se rompe la norma del contexto.

## Paso 7 – Restricciones éticas y jurídicas

### Actividad 7 – Definir el gate de admisibilidad

$$ADM(a) = Legal(a) \land Autonomia(a) \land Privacidad(a) \land Seguridad(a)$$

Condiciones obligatorias:

- **Legal:** cumple la Ley 19.496 y el Decreto 6/2021 (información clara del precio total, derecho de retracto, etc.).
- **Autonomía:** no usa patrones manipulativos (falsa escasez, cuentas regresivas falsas, casillas premarcadas, dificultar la cancelación).
- **Privacidad:** trata los datos personales según la normativa vigente, con consentimiento y finalidad clara.
- **Seguridad:** protege datos y pagos (HTTPS, PCI-DSS en pagos, gestión de vulnerabilidades).

| Alternativa | Legal | Autonomía | Privacidad | Seguridad | ADM |
|---|---|---|---|---|---|
| A ShopFast | PASS | **FAIL** | PASS | PASS | **FAIL** |
| B TrustCart | PASS | PASS | PASS | PASS | PASS |
| C LeanCommerce | PASS | PASS | PASS | PASS | PASS |
| D ProCommerce | PASS | PASS | PASS | PASS | PASS |

Con los datos de la guía ninguna queda INDETERMINATE, porque los gates vienen como True/False. En un caso real, si falta evidencia (por ejemplo, no hay auditoría de seguridad) la marcaría como INDETERMINATE y no la dejaría pasar hasta tenerla; en el código eso equivale a tratarla como FAIL.

## Paso 8 – Hecho, valoración y preferencia

En el modelo se distinguen tres cosas: el **hecho** (ej. TCO = \$35.000 es una medición), la **valoración** ($v_{TCO}(35.000) = 7.14$ es una transformación de valor) y la **preferencia** ($w_{TCO} = 0.15$ es una decisión dentro del modelo). En las celdas siguientes se ve cada una por separado.

## Paso 9 – Cargar los datos del caso

Los datos son los que entrega la guía. En un caso real cada dato debería tener valor + fuente + fecha + método + responsable.

In [1]:
# si no esta instalado: !pip install pulp
import pulp

alternativas = {
    "A_ShopFast": {
        "tco": 35000,
        "latencia": 250,
        "disponibilidad": 99.70,
        "PU": 4.4,
        "PEOU": 4.0,
        "mantenibilidad": 7.0,
        "persuasion_support": 10.0,
        # Gates no compensatorios
        "legal": True,
        "autonomia": False,
        "privacidad": True,
        "seguridad": True
    },
    "B_TrustCart": {
        "tco": 45000,
        "latencia": 300,
        "disponibilidad": 99.90,
        "PU": 4.5,
        "PEOU": 4.6,
        "mantenibilidad": 8.0,
        "persuasion_support": 7.5,
        "legal": True,
        "autonomia": True,
        "privacidad": True,
        "seguridad": True
    },
    "C_LeanCommerce": {
        "tco": 30000,
        "latencia": 420,
        "disponibilidad": 99.50,
        "PU": 4.0,
        "PEOU": 4.2,
        "mantenibilidad": 7.5,
        "persuasion_support": 6.5,
        "legal": True,
        "autonomia": True,
        "privacidad": True,
        "seguridad": True
    },
    "D_ProCommerce": {
        "tco": 52000,
        "latencia": 210,
        "disponibilidad": 99.95,
        "PU": 4.7,
        "PEOU": 4.1,
        "mantenibilidad": 8.5,
        "persuasion_support": 8.0,
        "legal": True,
        "autonomia": True,
        "privacidad": True,
        "seguridad": True
    }
}

print("Alternativas cargadas:", list(alternativas.keys()))

Alternativas cargadas: ['A_ShopFast', 'B_TrustCart', 'C_LeanCommerce', 'D_ProCommerce']


## Paso 10 – Funciones de valor (escala 0–10)

Anclas usadas (las de la guía): TCO \$25.000–\$60.000, latencia 150–600 ms, disponibilidad 99–100 %, TAM 1–5. Mantenibilidad y persuasión ya vienen en 0–10.

In [2]:
def valor_menor_mejor(x, minimo, maximo):
    valor = 10 * (maximo - x) / (maximo - minimo)
    return max(0, min(10, valor))


def valor_mayor_mejor(x, minimo, maximo):
    valor = 10 * (x - minimo) / (maximo - minimo)
    return max(0, min(10, valor))


def calcular_valores(alternativas):
    valores = {}
    for nombre, d in alternativas.items():
        valores[nombre] = {
            "costo": valor_menor_mejor(d["tco"], 25000, 60000),
            "rendimiento": valor_menor_mejor(d["latencia"], 150, 600),
            "disponibilidad": valor_mayor_mejor(d["disponibilidad"], 99.0, 100.0),
            "PU": valor_mayor_mejor(d["PU"], 1, 5),
            "PEOU": valor_mayor_mejor(d["PEOU"], 1, 5),
            "mantenibilidad": d["mantenibilidad"],
            "persuasion_support": d["persuasion_support"]
        }
    return valores


valores = calcular_valores(alternativas)

print("VALORES NORMALIZADOS (0-10)")
for a in valores:
    print(a, {c: round(v, 2) for c, v in valores[a].items()})

VALORES NORMALIZADOS (0-10)
A_ShopFast {'costo': 7.14, 'rendimiento': 7.78, 'disponibilidad': 7.0, 'PU': 8.5, 'PEOU': 7.5, 'mantenibilidad': 7.0, 'persuasion_support': 10.0}
B_TrustCart {'costo': 4.29, 'rendimiento': 6.67, 'disponibilidad': 9.0, 'PU': 8.75, 'PEOU': 9.0, 'mantenibilidad': 8.0, 'persuasion_support': 7.5}
C_LeanCommerce {'costo': 8.57, 'rendimiento': 4.0, 'disponibilidad': 5.0, 'PU': 7.5, 'PEOU': 8.0, 'mantenibilidad': 7.5, 'persuasion_support': 6.5}
D_ProCommerce {'costo': 2.29, 'rendimiento': 8.67, 'disponibilidad': 9.5, 'PU': 9.25, 'PEOU': 7.75, 'mantenibilidad': 8.5, 'persuasion_support': 8.0}


## Paso 11 – Definir qué puede compensarse (pesos)

Los pesos los asigné como analista a partir de lo que, según la guía, valora la organización. En un caso real deberían acordarse con los stakeholders.

| Criterio | Peso | Justificación | Rango ponderado |
|---|---|---|---|
| Costo / TCO | 0.15 | Importa a la organización, pero no queremos que lo más barato gane por sobre todo lo demás | \$25.000 – \$60.000 |
| Rendimiento | 0.10 | Todas las plataformas están en rangos aceptables (210–420 ms) | 150 – 600 ms |
| Disponibilidad | 0.10 | Todas están sobre 99,5 %, se le da peso moderado | 99 % – 100 % |
| PU | 0.20 | Si el cliente no encuentra útil la plataforma no compra; junto a PEOU es lo más importante (TAM) | Likert 1 – 5 |
| PEOU | 0.20 | La facilidad de uso afecta directamente que el cliente termine la compra | Likert 1 – 5 |
| Mantenibilidad | 0.10 | Importa al equipo TI a mediano plazo, pero el cliente no lo ve | 0 – 10 |
| Persuasión responsable | 0.15 | La organización quiere mecanismos de conversión, pero solo responsables (lo no responsable se controla en el gate) | 0 – 10 |
| **Total** | **1.00** | | |

In [3]:
pesos = {
    "costo": 0.15,
    "rendimiento": 0.10,
    "disponibilidad": 0.10,
    "PU": 0.20,
    "PEOU": 0.20,
    "mantenibilidad": 0.10,
    "persuasion_support": 0.15
}

assert abs(sum(pesos.values()) - 1.0) < 1e-9
print("Suma de pesos:", sum(pesos.values()))

Suma de pesos: 1.0


## Paso 12 – Scoring compensatorio

$$S(a) = \sum_{j=1}^{m} w_j \, v_j(a)$$

In [4]:
def calcular_scores(valores, pesos):
    score = {}
    for nombre in valores:
        score[nombre] = sum(
            pesos[c] * valores[nombre][c]
            for c in pesos
        )
    return score


score = calcular_scores(valores, pesos)

print("SCORING SIN GATES")
ranking = sorted(score.items(), key=lambda t: t[1], reverse=True)
for i, (nombre, s) in enumerate(ranking, 1):
    print(i, nombre, round(s, 3))

SCORING SIN GATES
1 A_ShopFast 7.949
2 B_TrustCart 7.685
3 D_ProCommerce 7.61
4 C_LeanCommerce 7.011


### Actividad 8 – Detectar el problema del scoring puro

1. **¿Qué alternativa obtiene el score más alto?** ShopFast, con 7.949.
2. **¿Esa alternativa respeta todos los valores no negociables?** No, falla en autonomía. El caso solo entrega el dato `autonomia: False` sin explicar el motivo; mi suposición es que su soporte de conversión (persuasión = 10) incluye prácticas como falsa escasez o temporizadores que no respetan la decisión del usuario.
3. **¿Sería correcto compensar una pérdida de autonomía con mejor costo y mayor conversión?** No. Si se permite compensar, se está aceptando que manipular a los clientes está bien siempre que salga más barato, lo que va contra los valores éticos y probablemente contra la ley del consumidor. ShopFast gana justamente por la persuasión (aporta 1.5 al score) y por el costo, que es lo que no debería poder "comprar" la autonomía. Por eso la autonomía se trata como veto y no como un criterio más con peso.

## Paso 13 – Gate no compensatorio

In [5]:
def es_admisible(d):
    return (
        d["legal"]
        and d["autonomia"]
        and d["privacidad"]
        and d["seguridad"]
    )


admisibles = {
    nombre: es_admisible(d)
    for nombre, d in alternativas.items()
}

print("ADMISIBILIDAD")
for nombre, ok in admisibles.items():
    fallas = [g for g in ["legal", "autonomia", "privacidad", "seguridad"]
              if not alternativas[nombre][g]]
    if ok:
        print(nombre, "PASS")
    else:
        print(nombre, "FAIL (falla:", ", ".join(fallas) + ")")

ADMISIBILIDAD
A_ShopFast FAIL (falla: autonomia)
B_TrustCart PASS
C_LeanCommerce PASS
D_ProCommerce PASS


## Paso 14 – PuLP

Se instala con `pip install pulp` (ya se importó en la primera celda de código). Se comprueba la versión:

In [6]:
print(pulp.__version__)

3.3.2


## Paso 15 – Formular el problema de optimización

Variable binaria $x_a = 1$ si se selecciona la alternativa $a$, 0 en otro caso.

$$\max Z = \sum_{a \in A} S(a)\,x_a \quad \text{s.a.} \quad \sum_{a \in A} x_a = 1, \qquad ADM(a) = FAIL \Rightarrow x_a = 0$$

In [7]:
def resolver_pulp(alternativas, score, admisibles):
    modelo = pulp.LpProblem("Seleccion_Plataforma_Ecommerce", pulp.LpMaximize)

    x = pulp.LpVariable.dicts("seleccionar", alternativas.keys(), cat="Binary")

    # Funcion objetivo
    modelo += pulp.lpSum(score[a] * x[a] for a in alternativas)

    # Elegir exactamente una plataforma
    modelo += pulp.lpSum(x[a] for a in alternativas) == 1

    # Vetos no compensatorios
    for a in alternativas:
        if not admisibles[a]:
            modelo += x[a] == 0

    modelo.solve(pulp.PULP_CBC_CMD(msg=False))
    return modelo, x

## Paso 16 – Resolver

In [8]:
modelo, x = resolver_pulp(alternativas, score, admisibles)

print("ESTADO:", pulp.LpStatus[modelo.status])

for a in alternativas:
    print(a, "score=", round(score[a], 3), "admisible=", admisibles[a], "x=", x[a].value())

for a in alternativas:
    if x[a].value() == 1:
        seleccionada = a
        print("\nSELECCIONADA:", a, "Score:", round(score[a], 3))

ESTADO: Optimal
A_ShopFast score= 7.949 admisible= False x= 0.0
B_TrustCart score= 7.685 admisible= True x= 1.0
C_LeanCommerce score= 7.011 admisible= True x= 0.0
D_ProCommerce score= 7.61 admisible= True x= 0.0

SELECCIONADA: B_TrustCart Score: 7.685


### Actividad 9 – Interpretar correctamente "Optimal"

**¿Por qué que PuLP muestre `Optimal` no demuestra que la alternativa sea la "mejor del mundo real"?**

Porque `Optimal` solo significa que PuLP encontró la mejor solución del **modelo** que le dimos. Ese modelo depende de la **evidencia** usada (TCO, latencia, encuestas TAM), que puede estar incompleta o mal medida; de las **preferencias** que nosotros definimos (pesos y anclas); y de las **restricciones** que decidimos incluir (elegir una y los gates). Si cambia cualquiera de esas cosas, el óptimo puede ser otra alternativa. Por lo tanto, `Optimal` demuestra que TrustCart es la mejor dado lo que pusimos en el modelo, no que sea la mejor en la realidad.

## Paso 17 – Trazabilidad del resultado

Permite reconstruir: dato → función de valor → peso → contribución → score.

In [9]:
campo_dato = {
    "costo": "tco", "rendimiento": "latencia",
    "disponibilidad": "disponibilidad", "PU": "PU", "PEOU": "PEOU",
    "mantenibilidad": "mantenibilidad",
    "persuasion_support": "persuasion_support"
}

def mostrar_detalle(alternativas, valores, score, admisibles):
    print("DETALLE")
    for a in alternativas:
        print("\n", a, "| Admisible:", admisibles[a], "| Score:", round(score[a], 3))
        for criterio in pesos:
            contribucion = pesos[criterio] * valores[a][criterio]
            print(f"   {criterio:<19} dato={alternativas[a][campo_dato[criterio]]:<8}"
                  f" valor={valores[a][criterio]:.3f}"
                  f" peso={pesos[criterio]:.2f}"
                  f" contribucion={contribucion:.3f}")

mostrar_detalle(alternativas, valores, score, admisibles)

DETALLE

 A_ShopFast | Admisible: False | Score: 7.949
   costo               dato=35000    valor=7.143 peso=0.15 contribucion=1.071
   rendimiento         dato=250      valor=7.778 peso=0.10 contribucion=0.778
   disponibilidad      dato=99.7     valor=7.000 peso=0.10 contribucion=0.700
   PU                  dato=4.4      valor=8.500 peso=0.20 contribucion=1.700
   PEOU                dato=4.0      valor=7.500 peso=0.20 contribucion=1.500
   mantenibilidad      dato=7.0      valor=7.000 peso=0.10 contribucion=0.700
   persuasion_support  dato=10.0     valor=10.000 peso=0.15 contribucion=1.500

 B_TrustCart | Admisible: True | Score: 7.685
   costo               dato=45000    valor=4.286 peso=0.15 contribucion=0.643
   rendimiento         dato=300      valor=6.667 peso=0.10 contribucion=0.667
   disponibilidad      dato=99.9     valor=9.000 peso=0.10 contribucion=0.900
   PU                  dato=4.5      valor=8.750 peso=0.20 contribucion=1.750
   PEOU                dato=4.6      va

## Paso 18 – Verificar y validar

### Verificación: ¿implementamos correctamente el modelo que diseñamos?

In [10]:
def verificar(alternativas, score, admisibles, x, seleccionada):
    suma_x = sum(x[a].value() for a in alternativas)
    print("Pesos suman 1:", abs(sum(pesos.values()) - 1) < 1e-9)
    print("Se elige exactamente una alternativa:", suma_x == 1)
    print("Inadmisibles con x=0:",
          all(x[a].value() == 0 for a in alternativas if not admisibles[a]))
    # el ganador de PuLP debe ser el mismo que el maximo entre admisibles
    mejor_admisible = max([a for a in alternativas if admisibles[a]], key=lambda a: score[a])
    print("PuLP coincide con el maximo admisible:", mejor_admisible == seleccionada)

verificar(alternativas, score, admisibles, x, seleccionada)

Pesos suman 1: True
Se elige exactamente una alternativa: True
Inadmisibles con x=0: True
PuLP coincide con el maximo admisible: True


### Actividad 10 – Verificación manual

Alternativa admisible elegida: **B TrustCart**.

- $v_{costo} = 10 \cdot (60000 - 45000)/(60000 - 25000) = 4.286$
- $v_{rend} = 10 \cdot (600 - 300)/(600 - 150) = 6.667$
- $v_{disp} = 10 \cdot (99.90 - 99)/(100 - 99) = 9.0$
- $v_{PU} = 10 \cdot (4.5 - 1)/(5 - 1) = 8.75$
- $v_{PEOU} = 10 \cdot (4.6 - 1)/(5 - 1) = 9.0$
- $v_{mant} = 8.0$, $v_{pers} = 7.5$

$$S(B) = 0.15 \cdot 4.286 + 0.10 \cdot 6.667 + 0.10 \cdot 9.0 + 0.20 \cdot 8.75 + 0.20 \cdot 9.0 + 0.10 \cdot 8.0 + 0.15 \cdot 7.5$$
$$= 0.643 + 0.667 + 0.900 + 1.750 + 1.800 + 0.800 + 1.125 = 7.685$$

Comparación con Python:

In [11]:
score_manual_B = 0.643 + 0.667 + 0.900 + 1.750 + 1.800 + 0.800 + 1.125
print("Score manual B:", round(score_manual_B, 3))
print("Score Python B:", round(score["B_TrustCart"], 3))

Score manual B: 7.685
Score Python B: 7.685


Coinciden (7.685), así que la implementación del score está bien.

### Validación: ¿el modelo representa suficientemente el problema real?

- **¿Faltan stakeholders?** Podrían faltar el área de atención al cliente o los proveedores de productos de la tienda, que también usan la plataforma.
- **¿Faltan criterios?** Sí, por ejemplo integración con medios de pago chilenos (Webpay), escalabilidad, accesibilidad, calidad del soporte y dependencia del proveedor.
- **¿Las escalas representan adecuadamente las preferencias?** Son lineales; en la realidad pasar de 99,5 % a 99,9 % de disponibilidad puede valer más que pasar de 99,0 % a 99,4 %.
- **¿TAM fue medido con usuarios apropiados?** No lo sabemos, la guía no indica cuántos usuarios respondieron ni quiénes eran. Es una debilidad.
- **¿Los gates éticos/jurídicos son pertinentes?** Sí, los cuatro no deberían negociarse, pero faltaría documentar con qué evidencia se dio PASS a cada uno.
- **¿Las anclas de normalización son defendibles?** Parecen razonables, pero no se indica de dónde salen (presupuesto, SLA, benchmarks). El ancla de TCO afecta bastante a ProCommerce, que es la más cara.

## Paso 19 – Análisis de sensibilidad

Se varía el peso de PEOU entre 0.10 y 0.30; el cambio se redistribuye proporcionalmente entre los demás criterios para que la suma siga siendo 1. "cambio?" se compara con el ganador usando los pesos base.

In [12]:
def ajustar_peso(base, criterio_objetivo, nuevo_peso):
    restantes = {k: v for k, v in base.items() if k != criterio_objetivo}
    suma_restantes = sum(restantes.values())

    nuevos = {}
    for k, v in base.items():
        if k == criterio_objetivo:
            nuevos[k] = nuevo_peso
        else:
            nuevos[k] = (v / suma_restantes) * (1 - nuevo_peso)
    return nuevos


def sensibilidad(alternativas, valores, admisibles, criterio, lista_pesos):
    score_base = calcular_scores(valores, pesos)
    candidatos = [a for a in alternativas if admisibles[a]]
    ganador_base = max(candidatos, key=lambda a: score_base[a])

    print("peso   ganador          score   2do lugar        diferencia  cambio?")
    for nuevo in lista_pesos:
        w = ajustar_peso(pesos, criterio, nuevo)
        scores_esc = calcular_scores(valores, w)

        ordenados = sorted(candidatos, key=lambda a: scores_esc[a], reverse=True)
        ganador = ordenados[0]
        segundo = ordenados[1]
        dif = scores_esc[ganador] - scores_esc[segundo]

        if ganador != ganador_base:
            cambio = "SI"
        else:
            cambio = "NO"

        print(f"{nuevo:.2f}   {ganador:<16} {scores_esc[ganador]:.3f}   "
              f"{segundo:<16} {dif:.3f}       {cambio}")


sensibilidad(alternativas, valores, admisibles, "PEOU", [0.10, 0.15, 0.20, 0.25, 0.30])

peso   ganador          score   2do lugar        diferencia  cambio?
0.10   D_ProCommerce    7.592   B_TrustCart      0.072       SI
0.15   B_TrustCart      7.602   D_ProCommerce    0.002       NO
0.20   B_TrustCart      7.685   D_ProCommerce    0.075       NO
0.25   B_TrustCart      7.767   D_ProCommerce    0.148       NO
0.30   B_TrustCart      7.849   D_ProCommerce    0.222       NO


### Actividad 11 – Robustez

| Peso PEOU | Alternativa ganadora | ¿Cambió? |
|---|---|---|
| 0.10 | D ProCommerce (7.592) | Sí |
| 0.15 | B TrustCart (7.602) | No |
| 0.20 | B TrustCart (7.685) | No |
| 0.25 | B TrustCart (7.767) | No |
| 0.30 | B TrustCart (7.849) | No |

**Clasificación: Sensible.** Al subir el peso de PEOU, TrustCart se mantiene y aumenta su ventaja, pero con 0.10 gana ProCommerce y con 0.15 la diferencia es solo 0.002 (prácticamente un empate). El cambio de ganador ocurre entre 0.10 y 0.15.

La razón es que TrustCart le gana a ProCommerce solo en costo y en PEOU (9.0 vs 7.75), mientras que ProCommerce es mejor en todo lo demás. Al bajar el peso de PEOU, esos otros criterios pesan más y ProCommerce pasa adelante. Sería recomendable probar también el peso del costo y conseguir mejor evidencia de PEOU, que es el criterio que define el resultado.

## Paso 20 – Argumento de credibilidad

### Actividad 12 – Redactar la conclusión

> "Se recomienda **B TrustCart** porque supera **los cuatro criterios no compensatorios (legal, autonomía, privacidad y seguridad)**. Entre las alternativas admisibles obtiene **el mayor score compensatorio (7.685)** bajo los pesos y funciones de valor definidos. La evidencia de aceptación proviene de **las mediciones TAM (PU = 4.5 y PEOU = 4.6 en escala Likert 1–5)**. La recomendación es **sensible** frente a variaciones de **el peso de PEOU (se mantiene entre 0.15 y 0.30, pero con 0.10 gana ProCommerce)**. Por ello, la conclusión se considera **justificable, pero no robusta** bajo las condiciones evaluadas."

Estructura Claim – Argument – Evidence:

| Claim | Evidencia | Estado |
|---|---|---|
| C0: TrustCart es una recomendación justificable | C1 a C5 | Se sostiene con reservas |
| C1: cumple criterios no compensatorios | E1: checklist ético/jurídico | Cumple |
| C2: aceptación tecnológica suficiente | E2: instrumento TAM | Cumple (no se conoce la muestra) |
| C3: maximiza el valor entre admisibles | E3: datos técnicos, E4: ejecución PuLP | Cumple |
| C4: el modelo fue verificado | Checklist y cálculo manual | Cumple |
| C5: robustez suficiente | E5: análisis de sensibilidad | Parcial |

ShopFast queda descartada aunque tenga el score más alto, porque falla en autonomía. Antes de decidir recomendaría una medición TAM más amplia comparando TrustCart y ProCommerce, ya que la decisión está entre esas dos.

## Desafío final – Quinta plataforma

Como la guía no le asigna nombre, la llamé **E_NovaShop**.

**1. Incorporar la alternativa**

In [13]:
alternativas_5 = dict(alternativas)
alternativas_5["E_NovaShop"] = {
    "tco": 41000,
    "latencia": 270,
    "disponibilidad": 99.85,
    "PU": 4.3,
    "PEOU": 4.7,
    "mantenibilidad": 7.8,
    "persuasion_support": 8.2,
    "legal": True,
    "autonomia": True,
    "privacidad": True,
    "seguridad": True
}

**2. Normalizar y 3. calcular score**

In [14]:
valores_5 = calcular_valores(alternativas_5)
score_5 = calcular_scores(valores_5, pesos)

print("Valores E_NovaShop:", {c: round(v, 3) for c, v in valores_5["E_NovaShop"].items()})
print("\nSCORING SIN GATES (5 alternativas)")
ranking_5 = sorted(score_5.items(), key=lambda t: t[1], reverse=True)
for i, (nombre, s) in enumerate(ranking_5, 1):
    print(i, nombre, round(s, 3))

Valores E_NovaShop: {'costo': 5.429, 'rendimiento': 7.333, 'disponibilidad': 8.5, 'PU': 8.25, 'PEOU': 9.25, 'mantenibilidad': 7.8, 'persuasion_support': 8.2}

SCORING SIN GATES (5 alternativas)
1 A_ShopFast 7.949
2 E_NovaShop 7.908
3 B_TrustCart 7.685
4 D_ProCommerce 7.61
5 C_LeanCommerce 7.011


**4. Comprobar admisibilidad**

In [15]:
admisibles_5 = {nombre: es_admisible(d) for nombre, d in alternativas_5.items()}
for nombre, ok in admisibles_5.items():
    print(nombre, "PASS" if ok else "FAIL")

A_ShopFast FAIL
B_TrustCart PASS
C_LeanCommerce PASS
D_ProCommerce PASS
E_NovaShop PASS


**5. Resolver con PuLP** (y verificación)

In [16]:
modelo_5, x_5 = resolver_pulp(alternativas_5, score_5, admisibles_5)
print("ESTADO:", pulp.LpStatus[modelo_5.status])

for a in alternativas_5:
    print(a, "score=", round(score_5[a], 3), "admisible=", admisibles_5[a], "x=", x_5[a].value())
    if x_5[a].value() == 1:
        seleccionada_5 = a

print("\nSELECCIONADA:", seleccionada_5, "Score:", round(score_5[seleccionada_5], 3))
print()
verificar(alternativas_5, score_5, admisibles_5, x_5, seleccionada_5)

ESTADO: Optimal
A_ShopFast score= 7.949 admisible= False x= 0.0
B_TrustCart score= 7.685 admisible= True x= 0.0
C_LeanCommerce score= 7.011 admisible= True x= 0.0
D_ProCommerce score= 7.61 admisible= True x= 0.0
E_NovaShop score= 7.908 admisible= True x= 1.0

SELECCIONADA: E_NovaShop Score: 7.908

Pesos suman 1: True
Se elige exactamente una alternativa: True
Inadmisibles con x=0: True
PuLP coincide con el maximo admisible: True


In [17]:
# trazabilidad solo de la nueva alternativa
mostrar_detalle({"E_NovaShop": alternativas_5["E_NovaShop"]}, valores_5, score_5, admisibles_5)

DETALLE

 E_NovaShop | Admisible: True | Score: 7.908
   costo               dato=41000    valor=5.429 peso=0.15 contribucion=0.814
   rendimiento         dato=270      valor=7.333 peso=0.10 contribucion=0.733
   disponibilidad      dato=99.85    valor=8.500 peso=0.10 contribucion=0.850
   PU                  dato=4.3      valor=8.250 peso=0.20 contribucion=1.650
   PEOU                dato=4.7      valor=9.250 peso=0.20 contribucion=1.850
   mantenibilidad      dato=7.8      valor=7.800 peso=0.10 contribucion=0.780
   persuasion_support  dato=8.2      valor=8.200 peso=0.15 contribucion=1.230


**6. Sensibilidad**

In [18]:
sensibilidad(alternativas_5, valores_5, admisibles_5, "PEOU", [0.10, 0.15, 0.20, 0.25, 0.30])

peso   ganador          score   2do lugar        diferencia  cambio?
0.10   E_NovaShop       7.740   D_ProCommerce    0.148       NO
0.15   E_NovaShop       7.824   B_TrustCart      0.221       NO
0.20   E_NovaShop       7.908   B_TrustCart      0.223       NO
0.25   E_NovaShop       7.992   B_TrustCart      0.225       NO
0.30   E_NovaShop       8.075   B_TrustCart      0.226       NO


En todo el rango (0.10 a 0.30) gana NovaShop, por lo que respecto a PEOU la recomendación es **robusta**.

**7. Conclusión argumentada**

> "Se recomienda **E NovaShop** porque supera **los cuatro criterios no compensatorios**. Entre las alternativas admisibles obtiene **el mayor score (7.908)** bajo los pesos y funciones de valor definidos. La evidencia de aceptación proviene de **las mediciones TAM (PU = 4.3, PEOU = 4.7)**. La recomendación es **robusta** frente a variaciones de **el peso de PEOU entre 0.10 y 0.30**. Por ello, la conclusión se considera **creíble** bajo las condiciones evaluadas."

NovaShop solo es la mejor en PEOU, pero no tiene puntajes bajos en ningún criterio (es equilibrada), y en una suma ponderada eso la favorece. Comparada con TrustCart le gana en costo, rendimiento, persuasión y PEOU, y pierde un poco en PU, disponibilidad y mantenibilidad. Solo se probó la sensibilidad de un peso; faltaría probar otros (costo, PU) para afirmar que es robusta en general.

## Preguntas de cierre

1. **¿Por qué MCDA no comienza necesariamente por los pesos?** Porque los pesos solo tienen sentido después de saber qué se valora y qué criterios representan esos valores. Si se parte por los pesos se termina ponderando lo que es fácil de medir y no lo que importa.
2. **¿Qué aporta Keeney a la construcción del modelo?** El Value-Focused Thinking: partir por los valores y objetivos y ver las alternativas como medios para lograrlos. En el taller se refleja en el árbol V0 → V1…V4.
3. **¿Qué diferencia existe entre la perspectiva de Roy y una búsqueda de "óptimo absoluto"?** Roy plantea que el modelo ayuda a decidir (decision aid) y se construye junto con los participantes; no existe una mejor alternativa objetiva independiente de los valores y el contexto.
4. **¿Por qué PU y PEOU requieren evidencia empírica?** Porque son percepciones de los usuarios; solo se pueden conocer preguntándoles con un instrumento válido, no suponiéndolas.
5. **¿Qué aporta Fogg y qué limitación ética debe añadirse a su uso?** Fogg explica cómo la tecnología puede influir en el comportamiento (recomendaciones, recordatorios, feedback). La limitación es que la influencia debe ser veraz, transparente y respetar la autonomía; si no, es manipulación.
6. **¿Cómo ayuda Value Sensitive Design a un e-commerce?** Obliga a preguntar qué valores afecta cada funcionalidad (privacidad, autonomía, confianza) y a diseñarla considerando eso desde el inicio, no después.
7. **¿Por qué privacidad puede tratarse como límite axiológico?** Porque una violación de privacidad no se arregla con menor costo o mejor rendimiento, y además es una obligación legal. Por eso va como veto.
8. **¿Qué diferencia existe entre un criterio compensatorio y un veto?** En el compensatorio una debilidad se puede compensar con otra fortaleza (trade-off aceptable). En el veto, si falla, la alternativa queda fuera sin importar su score.
9. **¿Qué demuestra realmente PuLP cuando entrega Optimal?** Que encontró la mejor solución del modelo formulado (scores, restricción y vetos), no que sea la mejor en la realidad.
10. **¿Qué diferencia existe entre verificar y validar?** Verificar es revisar que el modelo esté bien implementado (cálculos, código). Validar es revisar si el modelo representa bien el problema real (criterios, stakeholders, datos).
11. **¿Por qué un análisis de sensibilidad aumenta la credibilidad de la conclusión?** Porque muestra si la recomendación depende de un peso específico o se mantiene con cambios razonables. En el caso base mostró que la elección entre TrustCart y ProCommerce era frágil.
12. **¿Cómo se relacionan Claim, Argument y Evidence?** El claim es lo que se afirma (ej. "TrustCart es justificable"), la evidencia son los datos que lo apoyan (TAM, gates, PuLP, sensibilidad) y el argumento es el razonamiento que explica por qué esa evidencia sostiene el claim.